### PR TITLE
DRAFT: Allow user-driven cache invalidation 

### PR DESCRIPTION
**Pull request by bbenne10**
Starts work on allowing user-driven cache invalidation as discussed in #140.
Still needs documentation updates.
Opening as draft to discuss approach. 
Paging @dvzubarev and @Mic92 in particular.

### CODE REVIEW

**Comment by Mic92** [comment]
Current output from failing test:

```
AssertionError: 'renewed cache' not found in "direnv: loading ~/testenv/.envrc\ndirenv: using flake\n/home/runner/work/nix-direnv/nix-direnv/direnvrc:173: /tmp/tmpl_7745ne/testenv/.direnv/flake-profile-a5d5b61aa8a61b7d9d765e1daf971a9a578f1cfa.rc.tmp: No such file or directory\nmv: cannot stat '/tmp/tmpl_7745ne/testenv/.direnv/flake-profile-a5d5b61aa8a61b7d9d765e1daf971a9a578f1cfa.rc.tmp': No such file or directory\ndirenv: using cached dev shell\n/home/runner/work/nix-direnv/nix-direnv/direnvrc:185: /tmp/tmpl_7745ne/testenv/.direnv/flake-profile-a5d5b61aa8a61b7d9d765e1daf971a9a578f1cfa.rc: No such file or directory\n\x1b[31mdirenv: error command 'hello' not found on PATH '/nix/store/cwrb17acy0z8bv8kr7nrrhiz5nxvjfgx-bash-interactive-5.1-p12/bin:/nix/store/zd4k6kq962bgvlvh5x1kyl8sg7dpjz4y-shellcheck-0.8.0-bin/bin:/nix/store/lsy34q8izzzxxqbcl1balzmzb632r47q-direnv-2.30.3/bin:/nix/store/n9kf13nlax58wrq5573pahcdgk05fb8a-python3.9-mypy-0.931/bin:/nix/store/x24kxyqwqg2ln8kh9ky342kdcmhbng3h-python3-3.9.9/bin:/nix/store/30cbp6yraqafir4n4fswx6wc8ddlz6bw-python3.9-black-21.12b0/bin:/nix/store/2xzx6rf91lrf289rcy6yqy9amd50bcnf-python3.9-charset-normalizer-2.0.10/bin:/nix/store/d7gpbzckic9dbn818sj70npmbqvnp8pr-python3.9-cchardet-2.1.7/bin:/nix/store/zw6zazmrcnxjbw2fw3jipk131n9n9naa-python3.9-flake8-4.0.1/bin:/nix/store/l8gqlm2zrvs0z7dyyg0qkdbliahmzyj4-python3.9-pyflakes-2.4.0/bin:/nix/store/gc09x6syf49b7d8ggaxb75qg2a2b74bq-python3.9-pycodestyle-2.8.0/bin:/nix/store/pg971qqa30gmnj7j9wmqy07nhc9sdd2y-patchelf-0.14.3/bin:/nix/store/kb3x3n7f62ssrmmv4fy8shp4m9cfajl8-gcc-wrapper-10.3.0/bin:/nix/store/5kxrlhcbrfdjy4h0j51zc1q4hdgfnwsb-gcc-10.3.0/bin:/nix/store/wsa39231yijri5qn2r6bvmqw8sg55z12-glibc-2.33-108-bin/bin:/nix/store/b808b0pg675fww696g5c5l87h1r81c3y-coreutils-9.0/bin:/nix/store/lz5pb4y9z79lc65asdx6j0wiicm3p12q-binutils-wrapper-2.35.2/bin:/nix/store/aj74713k8yxxkl95agq5dqpcsxk4schi-binutils-2.35.2/bin:/nix/store/b808b0pg675fww696g5c5l87h1r81c3y-coreutils-9.0/bin:/nix/store/645v3545lcbx77wq7355rgdrgbhn5wx7-findutils-4.8.0/bin:/nix/store/pgmxmcad16axkgpl38fm2zzha25rdkqw-diffutils-3.8/bin:/nix/store/npxajdq5nc244mzhf52r0i4w0ynbprl3-gnused-4.8/bin:/nix/store/bafqr6466hd3r0kjf9lqjkm77bari839-gnugrep-3.7/bin:/nix/store/70xakzzda5k1qa5nm0d5gnz4zk38m8v0-gawk-5.1.1/bin:/nix/store/mly4acklgmf7fnlvcmczw4w0kkcy45w1-gnutar-1.34/bin:/nix/store/5f48000in35nhhp9xjf8vmdzsq7isrdv-gzip-1.11/bin:/nix/store/hwkv9ic0izrh4nw5z54fdkmpb4523kil-bzip2-1.0.6.0.2-bin/bin:/nix/store/472ba56b1j0z9k5gxgrlcn8mh17c1iaf-gnumake-4.3/bin:/nix/store/07j81a5xy8j1srvf57dl7lxsaihnmfac-bash-5.1-p12/bin:/nix/store/kcpgsvripkbyncj89iwhqgvvy3vcv8w7-patch-2.7.6/bin:/nix/store/jj023wxnhha23b8ikpyg48a2p3l0qb75-xz-5.2.5-bin/bin:/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/per-user/runner/profile/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/runner/.local/bin:/opt/pipx_bin:/home/runner/.cargo/bin:/home/runner/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'\x1b[0m\n"
```

**Comment by bbenne10** [comment]
I saw that. I think what MIGHT be happening is that I've inverted the directory cache invalidation logic? I'm not entirely sure. 

Currently running the tests locally (without realizing they do a gc sweep so I'm deleting a BUNCH of things) and I'll check on it.

**Comment by bbenne10** [comment]
That is exactly what happened. I'd returned the wrong value from the `_cache_is_valid` function. That's been fixed now.

File: direnvrc [hunk 1]
```diff
@@ -111,6 +111,24 @@ _nix_argsum_suffix() {
   fi
 }
 
+_cache_is_valid() {
+  if [ "$(type -t nix_direnv_cache_is_valid)"  == "function" ]; then
+    nix_direnv_cache_is_valid || return 1;
```
Comment (Mic92) [inline direnvrc:116]: Looks good. Can you document `nix_direnv_cache_is_valid` in the README as well?

**Comment by bbenne10** [comment]
I have also yet to investigate how this should or shouldn't work with `use_nix`. Do you have thoughts on that?

File: direnvrc (hunk 1 above)
Comment (bbenne10) [reply direnvrc]: Done in my last push.

**Comment by bbenne10** [comment]
I've got an invocation that *should* work with `use_nix`, but I've realized that the naming was out of whack with the expectations. (we'd return 0 if the cache was invalid)

I've inverted the if expression arms, which increased churn. But it's not too bad.

I'm sorry for thrashing CI publicly, but I simply cannot get the tests passing locally at all. 
I see a lot of "error <path>/.envrc is blocked. Run `direnv allow` to approve its content." (despite that being run during test setup)

**Comment by Mic92** [comment]
Looks good on the first look, can you, @dvzubarev also test if this fixes your usecase?

**Comment by bbenne10** [comment]
The test run implies it's very not okay. Something is wrong way around in the logic. I'm trying to look at it, but getting the tests passing locally is proving difficult. I am up on a deadline at work and am pretty busy otherwise. I'll try to fix this a bit this evening, but it may be this weekend before I clean it up fully.

I did get the tests running and failing in the same manner as on CI in a local docker image, so there's *some* progress.

**Comment by bbenne10** [comment]
Ugh. I was right the first time around with the return codes. I have that fixed locally. 
I am having a hard time debugging the last remaining failure.
It looks like `nix-collect-garbage` is cleaning up the `use nix` environment even after a gc root is added and I'm not entirely sure *why*. I'm going to push with what I have and see if this changes on CI too.

**Comment by bbenne10** [comment]
Unless there's some huge issue with this #148, I generally think that that PR is a better implementation of everything on display here. I am going to close this PR for the time being. If it is determined that #148 is an unsuitable replacement, we can reopen this.

**Pull request closed without merging**